docs/typescript.md

```diff
@@ -1,14 +1,597 @@
+
+```typescript
+function isFish(pet: Fish | Bird): pet is Fish {
+  return (pet as Fish).swim !== undefined;
+}
+```
+
+- Key remapping in mapped types uses the `as` keyword.
+
+```typescript
+type MappedTypeWithNewProperties<Type> = {
+  [Properties in keyof Type as NewKeyType]: Type[Properties];
+};
+```
+
+##### `as` is acceptable to use for typing data objects whose shape and contents are determined at runtime, externally, or through deserialization
+
+Preferably, this typing should be accompanied by runtime schema validation performed with type guards and unit tests.
+
+- e.g. The output of `JSON.parse()` or `await response.json()` for a known JSON input.
+- e.g. The type of a JSON file.
+
+<!-- TODO: Add example -->
+
+##### `as` may be acceptable to use in tests, for mocking or to exclude irrelevant but required properties from an input object
+
+It's recommended to provide accurate typing if there's any chance that omitting properties affects the accuracy of the test.
+
+If that is not the case, however, mocking only the properties needed in the test improves readability. It makes the intention and scope of the mock clear, and is more convenient to write, which can encourage test writing and improve test coverage.
+
+<!-- TODO: Add examples -->
+
+### Compiler Directives
+
+TypeScript provides several directive comments that can be used to suppress TypeScript compiler errors. Using these to ignore typing issues is dangerous and reduces the overall effectiveness of TypeScript.
+
+#### Acceptable usages of `@ts-expect-error`
+
+##### Use `@ts-expect-error` to force runtime execution of a branch for validation or testing
+
+Sometimes, there is a need to force a branch to execute at runtime for security or testing purposes, when that branch has correctly been inferred as being inaccessible by the TypeScript compiler.
+
+Suppressing compiler errors to avoid typing issues is usually dangerous because it results in a loss of information and safety. However, consciously using `@ts-expect-error` to add runtime checks represents an improvement in these aspects, and therefore is not discouraged.
+
+✅
+
+> **Error:** This comparison appears to be unintentional because the types '`0x${string}`' and '"__proto__"' have no overlap.ts(2367)
+
+```typescript
+exampleFunction(chainId: `0x${string}`) {
+    // @ts-expect-error Suppressing to perform runtime check
+    if (chainId === '__proto__') {
+      return;
+    }
+    ...
+}
+```
+
+##### `@ts-expect-error` may be acceptable to use in tests, to intentionally break features
+
+✅
+
+```typescript
+// @ts-expect-error Suppressing to test runtime error handling
+
+// @ts-expect-error Intentionally testing invalid state
+
+// @ts-expect-error We are intentionally passing bad input.
+```
+
+##### If accompanied by a TODO comment, `@ts-expect-error` is acceptable to use for marking errors that have clear plans of being resolved
+
+<!-- TODO: Add example -->
```

Comment:
Hmm. For this I was thinking of the cases where a TypeScript update results in new type errors, we could introduce a `@ts-expect-error` as a temporary fix so that we can resolve the type errors in a separate PR from the TypeScript update.

Maybe we can use the v4.4 update as the example here. It included changes to the `catch(error)` type that had a widespread impact, as you can see here: https://github.com/MetaMask/core/pull/795